providers/writing_a_new_provider.md

```diff
@@ -0,0 +1,569 @@
+      # attributes that you pass in here, and automatically add it to the `persister.flavors`
+      # inventory collection
+      persister.flavors.build(
+        # MIQ typically uses "ems_ref" as the unique reference for an inventory item
+        # in a provider.  Whatever you use must be guaranteed to not have a duplicate
+        # in the same provider instance as this value is also used to lookup related
+        # inventory from other collections.
+        :ems_ref => instance_type.id,
+        :name    => instance_type.name,
+        :cpus    => instance_type.n_cpus,
+        :memory  => instance_type.ram
+      )
+    end
+  end
+
+  def images
+    collector.images.each do |image|
+      persister.miq_templates.build(
+        :ems_ref         => image.id,
+        # The uid_ems field if you see it typically indicates a field that can be used
+        # to identify an inventory item across provider instances.  Most of the time
+        # for cloud providers this is the same as the ems_ref but not always.
+        # If your ems_ref looks like an integer ID then it probably isn't unique
+        # If it looks like a UUID then it probably is.
+        :uid_ems         => image.id,
+        :name            => image.name,
+        :location        => "unknown",
+        :raw_power_state => "never",
+        :template        => true,
+        :vendor          => "awesome_cloud"
+      )
+    end
+  end
+
+  def instances
+    collector.instances.each do |instance|
+      persister_vm = persister.vms.build(
+        :ems_ref         => instance.id,
+        :uid_ems         => instance.id,
+        :name            => instance.name,
+        :location        => instance.availability_zone,
+        :raw_power_state => instance.power_state,
+        :vendor          => "awesome_cloud",
+        # This is where things get interesting.  A "Lazy Reference" is our way of
+        # declaring a relationship to another table.  The result of this lazy_find
+        # after save_inventory has completed will be a foreign-key to that table
+        #
+        # It is critically important that you use this instead of either:
+        # 1. Direct database query like: Flavor.find_by(:ems_ref => instance.flavor)
+        #   because this won't work with flavors that are being created in this refresh
+        # 2. Using data set in the parser like: persitser.flavors.data.detect { |f| f[:ems_ref] == instance.flavor }
+        #   because this introduces an ordered dependency (flavors have to be parsed
+        #   before instances) and it is possible to introduce a dependency cycle and
+        #   also makes future targeted refresh much harder (where the flavors might not be
+        #   present in the collected data).
+        :flavor          => persister.flavors.lazy_find(instance.flavor)
+      )
+    end
+  end
+end
+```
+
+You'll have to add your vendor name to the core `VmOrTemplate` `VENDOR_TYPES` in order for the VMs to be saved.
```

Comment:
NOTE: we have a checklist item for this https://github.com/ManageIQ/manageiq/issues/19440